File: app/robust/merge_list.py
# coding=utf-8
# author: Xiguang Liu<g10guang@foxmail.com>
# 2018-04-27 17:28
# 题目描述：https://www.nowcoder.com/practice/d8b6b4358f774294a89de2a6ac4d9337?tpId=13&tqId=11169&tPage=1&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking

# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    # 返回合并后列表
    def Merge(self, pHead1, pHead2):
        if pHead2 is None:
            return pHead1
        if pHead1 is None:
            return pHead2
        p, q = pHead1, pHead2
        newHead = None
        last = None
        while p is not None and q is not None:
            if p.val <= q.val:
                n = p
                p = p.next
            else:
                n = q
                q = q.next
            if newHead is None:
                newHead = n
                last = newHead
            else:
                last.next = n
                last = n
        r = p or q
        last.next = r
        return newHead

File: app/robust/merge_list.go
package main

/*
 * type ListNode struct{
 *   Val int
 *   Next *ListNode
 * }
 */

/**
 *
 * @param pHead1 ListNode类
 * @param pHead2 ListNode类
 * @return ListNode类
 */
//  https://www.nowcoder.com/practice/d8b6b4358f774294a89de2a6ac4d9337?tpId=13&tqId=11169&tPage=1&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking
func Merge(pHead1 *ListNode, pHead2 *ListNode) *ListNode {
	if pHead1 == nil {
		return pHead2
	}
	if pHead2 == nil {
		return pHead1
	}
	p := pHead1
	q := pHead2
	var root, t *ListNode
	for p != nil && q != nil {
		var n *ListNode
		if p.Val < q.Val {
			n = p
			p = p.Next
		} else {
			n = q
			q = q.Next
		}
		n.Next = nil
		if root == nil {
			root = n
			t = n
		} else {
			t.Next = n
			t = n
		}
	}
	if p != nil {
		t.Next = p
	}
	if q != nil {
		t.Next = q
	}
	return root
}
